multisig/src/lib.rs

```diff
@@ -201,7 +201,18 @@ impl MultiSigContract {
                     }
                 }
                 MultiSigRequestAction::DeleteKey { public_key } => {
-                    promise.delete_key(public_key.into())
+                    let pk:PublicKey = public_key.into();
+                    // delete outstanding requests by public_key
+                    let request_ids = self.list_request_ids();
+                    for request_id in request_ids {
+                        let request_with_signer = self.requests.get(&request_id).expect("No such request");
```

Comment:
You can iterate on values instead and remember indices that has to be removed.
Then you can collect keys by index (request Ids) and only then remove such requests in a loop.

This will be more efficient because you will skip a few reads required to first get the request_id for items that don't match the signer_pk.

For this you can use `keys_as_vector` and `values_as_vector` methods of `UnorderedMap`